src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -81,33 +82,33 @@ private static (IReadOnlyDictionary<FieldDeclaration, InputModelProperty> Fields
         {
             var fieldsToInputs = new Dictionary<FieldDeclaration, InputModelProperty>();
             var publicParameters = new List<Parameter>();
-            var serializatoinParameters = new List<Parameter>();
+            var serializationParameters = new List<Parameter>();
             var parametersToFields = new Dictionary<string, FieldDeclaration>();
 
             foreach (var inputModelProperty in inputModel.Properties)
             {
                 var propertyIsCollection = inputModelProperty.Type is InputDictionaryType or InputListType;
                 var propertyIsRequiredInNonRoundTripModel = (inputModel.Usage is InputModelTypeUsage.Input or InputModelTypeUsage.Output) && inputModelProperty.IsRequired;
-                var propertyIsReadOnly = inputModelProperty.IsReadOnly || propertyIsCollection || propertyIsRequiredInNonRoundTripModel;
+                var propertyIsOptionalInOutputModel = (inputModel.Usage is InputModelTypeUsage.Output) && !inputModelProperty.IsRequired;
+                var propertyIsReadOnly = inputModelProperty.IsReadOnly || propertyIsCollection || propertyIsRequiredInNonRoundTripModel || propertyIsOptionalInOutputModel;
 
                 var fieldModifiers = propertyIsReadOnly ? Public | ReadOnly : Public;
                 var fieldType = GetDefaultPropertyType(inputModel.Usage, inputModelProperty, typeFactory);
 
-                var field = new FieldDeclaration($"{inputModelProperty.Description}", fieldModifiers, fieldType, inputModelProperty.Name.FirstCharToUpperCase(), writeAsProperty: true);
+                var field = new FieldDeclaration($"{inputModelProperty.Description}", fieldModifiers, fieldType, inputModelProperty.Name.FirstCharToUpperCase(),
+                    isRequired: inputModelProperty.IsRequired, writeAsProperty: true);
```

Comment:
Is there a difference between read-only and required field?

Reply:
> Is there a difference between read-only and required field?

@AlexanderSher There is. See the [example](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-model-readonly-property/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.cs) regarding read-only required/optional fields.
A read-only value type field is different when it's required or not. So we need a flag to indicate whether the field is required or not.
- required: defined as `T`
- optional: defined as `T?`